### PR TITLE
NEO-1403: Fix 'Selected' state of options in React Select

### PR DESCRIPTION
**Pull request by mrazian85**
[Link to updated components in Deploy Preview](https://deploy-preview-136--neo-react-library-storybook.netlify.app/?path=/story/components-select--basic-selects)

**Before tagging the team for a review, I have done the following:**
- [x] reviewed my code changes
- [x] ensured that all tests pass
- [x] updated the link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [x] tagged Matt if any visual changes have occurred
- [x] done an accessibility check on my work

The work in this PR is for [NEO-1403](https://jira.forge.avaya.com/browse/NEO-1403), which concerns the behaviour whereby the `aria-selected` attribute of options in the Select is being toggled on hover.

However, while working to resolve the above, I discovered some deeper issues with the way in which the custom behaviour we have implemented for the Select is overriding or otherwise colliding with the behaviour of the `Downshift` library which comprises the basis of the Select Component behaviour.

It took me quite some time to identify and try to fix these issues, and so I want to open a draft PR as sort of a 'sanity check' to get other eyes on this and ensure that my approach is sound. Further explanation is provided via comments here in Github where appropriate/necessary.

Will tag @matthewAvaya for further visual/a11y testing once logic is agreed.

### CODE REVIEW

File: src/components/Select/InternalComponents/InternalSelectOption.tsx [hunk 1]
```diff
@@ -56,13 +55,12 @@ export const InternalSelectOption = ({
   logger.debug(itemProps);
 
   return multiple ? (
-    <div className="neo-input-group">
```
Comment (mrazian85) [inline src/components/Select/InternalComponents/InternalSelectOption.tsx:59]: This `div` was not necessary for styling or implementation, so I have removed it.

File: src/components/Select/utils/useDownshift.ts [hunk 2]
```diff
@@ -281,7 +280,6 @@ const DownshiftWithMultipleSelectProps = (
           return {
             ...changes,
             highlightedIndex: state.highlightedIndex,
-            inputValue: "",
```
Comment (mrazian85) [inline src/components/Select/utils/useDownshift.ts:284]: Doesn't seem to do or affect anything, so removed

File: src/components/Select/InternalComponents/OptionWithCheckbox.tsx [hunk 3]
```diff
@@ -1,60 +1,65 @@
...
-        switch (event.key) {
-          case Keys.SPACE:
-          case Keys.ENTER:
-            setState(!state);
-            break;
-        }
-      }
-    };
     return (
-      <option
```
Comment (mrazian85) [inline src/components/Select/InternalComponents/OptionWithCheckbox.tsx:42]: Using an `option` here was not triggering the correct behaviour in `Downshift`. I have replaced it with a `li` as per [their examples in the docs](https://www.downshift-js.com/use-select)

File: src/components/Select/InternalComponents/OptionWithCheckbox_shim.css [hunk 4]
```diff
@@ -62,6 +62,10 @@
   cursor: not-allowed;
 }
 
+.neo-option.neo-option--focused {
```
Comment (mrazian85) [inline src/components/Select/InternalComponents/OptionWithCheckbox_shim.css:65]: Adding a separate style to distinguish 'selected' state from 'focused' state as these two are visually different

File: src/components/Select/InternalComponents/OptionWithCheckbox.stories.tsx [hunk 5]
```diff
@@ -1,18 +1,28 @@
...
+import { OptionProps, OptionWithCheckbox } from "./OptionWithCheckbox";
 
 export default {
   title: "Components/Select/OptionWithCheckbox",
   component: OptionWithCheckbox,
 } as Meta<OptionProps>;
 
 export const Default = () => {
-  return <OptionWithCheckbox defaultChecked>Option</OptionWithCheckbox>;
+  return (
```
Comment (mrazian85) [inline src/components/Select/InternalComponents/OptionWithCheckbox.stories.tsx:11]: Tweaking test to ensure `li` is wrapped in `ul` for correct a11y

File: src/components/Select/InternalComponents/OptionWithCheckbox.test.jsx [hunk 6]
```diff
@@ -1,18 +1,32 @@
 import { composeStories } from "@storybook/testing-react";
 import { render } from "@testing-library/react";
 import { axe } from "jest-axe";
+import { vi } from "vitest";
 
+import { SelectContext } from "../utils/SelectContext";
 import { OptionWithCheckbox } from "./OptionWithCheckbox";
 import * as OptionStories from "./OptionWithCheckbox.stories";
 
+const wrapper = ({ children }) => (
```
Comment (mrazian85) [inline src/components/Select/InternalComponents/OptionWithCheckbox.test.jsx:10]: This mocks the values provided by the `Downshift` library for the purpose of these tests

File: src/components/Select/InternalComponents/OptionWithCheckbox.tsx [hunk 7]
```diff
@@ -1,60 +1,65 @@
...
+    const itemProps = getItemProps({
+      item: optionSelf,
+      index,
+      "aria-selected": optionSelf.value
+        ? selectedItemsValues.includes(optionSelf.value)
+        : false,
+      "aria-describedby": helperText && helperId,
+      onClick: (event) => {
+        // Missing type definitions in Downshift, see https://github.com/downshift-js/downshift/issues/734
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (mrazian85) [inline src/components/Select/InternalComponents/OptionWithCheckbox.tsx:38]: Lines 39 and 44 ensure that the `disabled` prop functions correctly on a `li`

**Review by yangAtSpoken — COMMENTED** [review]
Looks good.  Could you fix the OptionWithCheckbox stories?

File: src/components/Select/InternalComponents/OptionWithCheckbox.tsx [hunk 8]
```diff
@@ -1,60 +1,65 @@
...
         className={clsx(
           "neo-option",
-          state && "neo-option--selected",
-          disabled && "neo-option--disabled"
+          selectedItemsValues.includes(optionSelf.value) &&
+            "neo-option--selected",
+          disabled && "neo-option--disabled",
+          index === highlightedIndex && "neo-option--focused"
         )}
-        onClick={() => setState(!state)}
```
[the remark this answers is not in the document]
Comment (enrique-prado) [reply src/components/Select/InternalComponents/OptionWithCheckbox.tsx:50]: Sounds reasonable. If you haven't done it already, I would just manually test different variants to make sure it works as expected.

**Comment by mrazian85** [comment]
> Looks good. Could you fix the OptionWithCheckbox stories?

Ouf, good catch @yangAtSpoken. I knew I would forget something. I have fixed it now, although I am not sure of the implementation as the `OptionWithCheckbox` now depends on the information provided by `Downshift`. Please lmk your thoughts 🙏

**Review by enrique-prado — APPROVED** [review]
Just the one comment about manually testing different variants.

**Comment by matthewAvaya** [comment]
**Basic Selects**
- When selecting multiple items, the checkbox from the dropdown menu should be filled. Instead, it is in an unselected state (applies to all multiple select versions)
- Looks like the single select jumps up by 1px when it's on focus.
**Creatable**
- Multiple select is not showing dropdown results while I'm typing a search word. This occurs on select that are searchable. 
Required In Form
- Is there a reason the 'Multiple Foods Selection' has a gray background? It looks disabled. 
**Load Options**
- When the select is loading, there's a white background on top and bottom part of the component. 
Screen Shot 2023-02-06 at 10 47 44 AM
**All Select components**
- Change down chevron icon to be facing up when a dropdown is used.
**Accessibility**
- Still getting selected items read twice by the screen reader.
- Screen reader is picking up some weird text, see screenshot. 
Screen Shot 2023-02-06 at 10 54 08 AM

**Comment by mrazian85** [comment]
@matthewAvaya, as discussed, please find comments or otherwise the relevant JIRA ticket numbers for the issues listed. I also created [NEO-1452](https://jira.forge.avaya.com/browse/NEO-1451) to track the issue where disabled Selects are interactive.

> **Basic Selects**
> 
> * When selecting multiple items, the checkbox from the dropdown menu should be filled. Instead, it is in an unselected state (applies to all multiple select versions) **(FIXED)**
> * Looks like the single select jumps up by 1px when it's on focus. ([NEO-1450](https://jira.forge.avaya.com/browse/NEO-1450))
>   **Creatable**
> * Multiple select is not showing dropdown results while I'm typing a search word. This occurs on select that are searchable. ([NEO-1450](https://jira.forge.avaya.com/browse/NEO-1451))
>   Required In Form
> * Is there a reason the 'Multiple Foods Selection' has a gray background? It looks disabled. (**CAN'T REPRODUCE**)
>   **Load Options**
> * When the select is loading, there's a white background on top and bottom part of the component. ([NEO-1454](https://jira.forge.avaya.com/browse/NEO-1454))
> 
> Screen Shot 2023-02-06 at 10 47 44 AM
> 
> **All Select components** - Change down chevron icon to be facing up when a dropdown is used. ([NEO-1456](https://jira.forge.avaya.com/browse/NEO-1456)) **Accessibility** - Still getting selected items read twice by the screen reader. ([NEO-1389](https://jira.forge.avaya.com/browse/NEO-1389) in current Sprint) Screen reader is picking up some weird text, see screenshot. (**CAN'T REPRODUCE**)
> Screen Shot 2023-02-06 at 10 54 08 AM

**Review by yangAtSpoken — APPROVED** [review]

**Pull request merged**